src/twisted/internet/defer.py

```diff
@@ -1456,44 +1540,64 @@ def _inlineCallbacks(result, g, status):
             # The traceback starts in this frame (the one for
             # _inlineCallbacks); the next one down should be the application
             # code.
-            appCodeTrace = exc_info()[2].tb_next
+            excInfo = exc_info()
+            assert excInfo is not None
+            traceback = excInfo[2]
+            assert traceback is not None
+            appCodeTrace = traceback.tb_next
+            assert appCodeTrace is not None
 
             # The contextvars backport and our no-op shim add an extra frame.
             if version_info < (3, 7):
                 appCodeTrace = appCodeTrace.tb_next
 
+            # type note: a number of ignore[union-attr] cases follow…
+            # It seems mypy isn't processing "assert is not None" as expected
+            # in some cases here.
+
```

Comment:
I feel like I must be doing something wrong here but I don't see it

Reply:
what does `reveal_type(appCodeTrace)` say here?

Reply:
might also be nicer with https://docs.python.org/3/library/traceback.html#traceback.walk_tb

Reply:
Aha! `appCodeTrace` gets redefined and I wasn't re-checking it.